GSOC-23: Roles and User Modules

chaitak-gorai opened the pull request:
## Description
This PR is for the Google Summer of Code Project : **Migrate EHR to Laravel: Users and Roles Module** .

## Workflow
- Create Roles Modules with Access Controls with permissions.
    - Identify ACL permissions from legacy codebase.
    - Create permissions
    - Create Role with those permissions
- Create User Module
- Integrate Users and Roles Module

*All the commits will be added to this for the final merge to the codebase

chaitak-gorai committed:
fixed head for mobile devices

chaitak-gorai committed:
Added icon template for svg icons

chaitak-gorai committed:
✨ Drawer Navigation Menu

chaitak-gorai committed:
Fixed chevron icon color

chaitak-gorai committed:
🐛 fixed the backdrop bug

chaitak-gorai committed:
added all nav menu

chaitak-gorai committed:
Added L2 submenus

chaitak-gorai committed:
Added L2 submenus for small devices

chaitak-gorai committed:
Roles interface with add role

chaitak-gorai commented:
## Added Roles UI with functionalities of creating and fetching roles


- Created `RolesController` for the backend logic
- Created the UI for the Roles Page with `Add Role` Form and `Existing Roles` Section in `Roles.vue`
- Added the routes in `web.php`

## Screenshot 

![image](https://github.com/LibreHealthIO/lh-ehr-laravel/assets/77141674/5a1c118e-b790-4953-b514-7c6afcc99677)

muarachmann commented on app/Http/Controllers/Dashboard/RolesController.php:
```diff
@@ -0,0 +1,68 @@
+        ]);
+
+        if ($validator->fails()) {
+            return response()->json($validator->errors(), 422);
```
No need to do this. Inertia handles this if validation fails. 
You should consider using a request form here rather

muarachmann commented on resources/app/pages/Roles.vue:
```diff
@@ -0,0 +1,171 @@
+        addRole() {
+            axios
+                .post("/dashboard/roles", {
+                    name: this.name,
```
Use ziggy for routing... It's in this project

muarachmann commented on routes/web.php:
```diff
@@ -109,7 +110,9 @@ function () {
-
+            Route::get('/roles', [RolesController::class, 'index']);
+            Route::post('/roles', [RolesController::class, 'store']);
+            Route::get('/allRoles', [RolesController::class, 'getRoles']);
```
use slug naming for routing e.g `all-roles`

chaitak-gorai replied:
Ok, I ll be updating these changes.

muarachmann reviewed (commented):
@chaitak-gorai left some reviews

chaitak-gorai committed:
🐛 Minor fixes: used ziggy

chaitak-gorai committed:
Added Laratrust Seeder for Roles

chaitak-gorai commented:
## Update: Added Laratrus Seeder for Roles 

@muarachmann, I have updated the changes you suggested in the earlier commit.
In a later commit, I added the Laratrust Seeder, which will assign 'c,r,u,d' permissions to each module's role.

In `config/laratrust_seeder.php` I have done this :
- Super Admin CRUD for all modules
- Admin is the same as super admin except for no update and deletes in administration
- lr-user: read for all
- accounting: fees and reports CRUD
- clinicians: CRUD for patient-related modules else RU
- Emergency login: Read for all and CRUD for the patient(keeping in mind the emergency_login will be done by a patient_
- front office: generally RU
- physicians: generally CRUD except for the administration
- supper student: RU

Can you please REVIEW these access and suggest to me the changes we need to make??

muarachmann commented on app/Http/Controllers/Dashboard/RolesController.php:
```diff
@@ -0,0 +1,59 @@
+     // Get the permissions associated with each role
+
+
+     // Embed the permissions into the roles
```
Remove unwanted comments if not used.

chaitak-gorai replied:
Fixed it.

muarachmann commented on config/laratrust_seeder.php:
```diff
@@ -0,0 +1,141 @@
+            'qa-measures' => 'c,r,u,d',
+            'help' => 'r',
+        ],
+        'lh-user' => [
```
Let's call this role `users`

chaitak-gorai replied:
Ok, I have updated it in the recent commit.

muarachmann commented:
I will review the roles/module in details later this week

chaitak-gorai committed:
Minor updates

chaitak-gorai committed:
Role UI update and Permission Data fetched in form

chaitak-gorai committed:
Seeder Temporary bug fix & Scrollable Dropdown UI

chaitak-gorai commented:
## Update: 

@muarachmann, I have updated some changes here I m listing for your review.
- `Help Required` Removing old roles completely giving too many redirects. I can't find anywhere the old roles are used. If there is, can you please help me out where these are used?
- Temporarily, I have solved it by keeping both new and old roles. Commented the truncate function in `LaratrustSeeder.php`
- The Existing Roles section is too long so applied `scroll` there.
- Also made the dropdown for the permissions scrollable too to fit the large list.
 
Screenshot of update Roles.vue UI
![image](https://github.com/LibreHealthIO/lh-ehr-laravel/assets/77141674/981f7f61-6296-4edb-8978-336f05879354)


Can you please REVIEW these changes and suggest to me the changes we need to make??

chaitak-gorai committed:
🐛 Seeder updated & Bug solved

chaitak-gorai committed:
Seeder updated with sub modules

chaitak-gorai committed:
Customized Role Creation

chaitak-gorai committed:
Users in Roles UI

chaitak-gorai committed:
Roles Details Page

muarachmann commented on app/Http/Controllers/Dashboard/RolesController.php:
```diff
@@ -0,0 +1,85 @@
+        $users = $role->users()->get();
+        $temp = [
+            'id' => $role->id,
+            'name' => $role->name,
```
Indent code ....

muarachmann commented on resources/app/pages/Roles.vue:
```diff
@@ -0,0 +1,268 @@
+                .post(this.route("dashboard.roles.store"), {
+                    name: this.roleForm.name,
+                    display_name: this.roleForm.display_name,
+                    description: this.roleForm.description,
```
Use inertia here to add a role

muarachmann commented on app/Http/Controllers/Dashboard/RolesController.php:
```diff
@@ -0,0 +1,85 @@
+            ]
+        );
+    }
+    public function details($roleId): Response
```
indentation, spacing and clean code

chaitak-gorai committed:
Minor updates

chaitak-gorai committed:
Add User Form UI

chaitak-gorai committed:
Add User Form UI

chaitak-gorai commented:
## Update:

@muarachmann, I have updated your suggested reviews. Also, I have created the `Add Users` UI form page with the fields;
I have added all necessary form fields. I have a doubt whether the name can be in a single `Full Name` field or in parts.
Please review it and suggest changes if required.

I ll then implement the functions

![image](https://github.com/LibreHealthIO/lh-ehr-laravel/assets/77141674/4283a838-75db-4878-aa4c-63fd8d108006)

tmccormi commented:
We should stick with separate fields.

… But I find it an interesting thought experiment to consider the
implications of one name field as better internationally.   Search is fast
and easy these days, so finding patients would not be an issue.

Sharing data externally would; in the USA, as healthcare businesses are all
stuck in old models.

Tony

On Fri, Jul 7, 2023 at 5:56 PM Chaitak Gorai ***@***.***>
wrote:

> Update:
>
> @muarachmann <https://github.com/muarachmann>, I have updated your
> suggested reviews. Also, I have created the Add Users UI form page with
> the fields;
> I have added all necessary form fields. I have a doubt whether the name
> can be in a single Full Name field or in parts.
> Please review it and suggest changes if required.
>
> I ll then implement the functions
>
> [image: image]
> <https://user-images.githubusercontent.com/77141674/251884447-4283a838-75db-4878-aa4c-63fd8d108006.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LibreHealthIO/lh-ehr-laravel/pull/43#issuecomment-1626399181>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACFZC324C54XWM3WP4KX4LXPCV2BANCNFSM6AAAAAAZA67IAQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
https://thinkstorm.net
503-330-2239

tmccormi commented:
The users name as a single field would work , if we ONLY used it for user
identification purpose, however we use the provider’s user account on
external forms that require separate fields, including title.

On Sat, Jul 8, 2023 at 9:05 AM Tony McCormick ***@***.***>
wrote:

> We should stick with separate fields.
>
> … But I find it an interesting thought experiment to consider the
> implications of one name field as better internationally.   Search is fast
> and easy these days, so finding patients would not be an issue.
>
> Sharing data externally would; in the USA, as healthcare businesses are
> all stuck in old models.
>
> Tony
>
> On Fri, Jul 7, 2023 at 5:56 PM Chaitak Gorai ***@***.***>
> wrote:
>
>> Update:
>>
>> @muarachmann <https://github.com/muarachmann>, I have updated your
>> suggested reviews. Also, I have created the Add Users UI form page with
>> the fields;
>> I have added all necessary form fields. I have a doubt whether the name
>> can be in a single Full Name field or in parts.
>> Please review it and suggest changes if required.
>>
>> I ll then implement the functions
>>
>> [image: image]
>> <https://user-images.githubusercontent.com/77141674/251884447-4283a838-75db-4878-aa4c-63fd8d108006.png>
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/LibreHealthIO/lh-ehr-laravel/pull/43#issuecomment-1626399181>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AACFZC324C54XWM3WP4KX4LXPCV2BANCNFSM6AAAAAAZA67IAQ>
>> .
>> You are receiving this because you are subscribed to this thread.Message
>> ID: ***@***.***>
>>
> --
> https://thinkstorm.net
> 503-330-2239
>
> --
https://thinkstorm.net
503-330-2239

chaitak-gorai commented:
> The users name as a single field would work , if we ONLY used it for user identification purpose, however we use the provider’s user account on external forms that require separate fields, including title. On Sat, Jul 8, 2023 at 9:05 AM Tony McCormick ***@***.***> wrote:
> We should stick with separate fields. … But I find it an interesting thought experiment to consider the implications of one name field as better internationally. Search is fast and easy these days, so finding patients would not be an issue. Sharing data externally would; in the USA, as healthcare businesses are all stuck in old models. Tony On Fri, Jul 7, 2023 at 5:56 PM Chaitak Gorai ***@***.***> wrote: > Update: > > @muarachmann <https://github.com/muarachmann>, I have updated your > suggested reviews. Also, I have created the Add Users UI form page with > the fields; > I have added all necessary form fields. I have a doubt whether the name > can be in a single Full Name field or in parts. > Please review it and suggest changes if required. > > I ll then implement the functions > > [image: image] > <https://user-images.githubusercontent.com/77141674/251884447-4283a838-75db-4878-aa4c-63fd8d108006.png> > > — > Reply to this email directly, view it on GitHub > <[#43 (comment)](https://github.com/LibreHealthIO/lh-ehr-laravel/pull/43#issuecomment-1626399181)>, > or unsubscribe > <https://github.com/notifications/unsubscribe-auth/AACFZC324C54XWM3WP4KX4LXPCV2BANCNFSM6AAAAAAZA67IAQ> > . > You are receiving this because you are subscribed to this thread.Message > ID: ***@***.***> > -- https://thinkstorm.net 503-330-2239 --
> https://thinkstorm.net 503-330-2239

so as of now can we move forward with a single user name field? I think it will be an efficient way to store the name.

tmccormi commented:
No, The provider data is used on claim forms that require separate fields.
Wish it was otherwise.  Mua, what do you think?

Tony

On Sat, Jul 8, 2023 at 9:16 AM Chaitak Gorai ***@***.***>
wrote:

> The users name as a single field would work , if we ONLY used it for user
> identification purpose, however we use the provider’s user account on
> external forms that require separate fields, including title. On Sat, Jul
> 8, 2023 at 9:05 AM Tony McCormick *@*.
> *> wrote: We should stick with separate fields. … But I find it an
> interesting thought experiment to consider the implications of one name
> field as better internationally. Search is fast and easy these days, so
> finding patients would not be an issue. Sharing data externally would; in
> the USA, as healthcare businesses are all stuck in old models. Tony On Fri,
> Jul 7, 2023 at 5:56 PM Chaitak Gorai @.*> wrote: > Update: > >
> @muarachmann <https://github.com/muarachmann>
> https://github.com/muarachmann, I have updated your > suggested reviews.
> Also, I have created the Add Users UI form page with > the fields; > I have
> added all necessary form fields. I have a doubt whether the name > can be
> in a single Full Name field or in parts. > Please review it and suggest
> changes if required. > > I ll then implement the functions > > [image:
> image] >
> https://user-images.githubusercontent.com/77141674/251884447-4283a838-75db-4878-aa4c-63fd8d108006.png
> > > — > Reply to this email directly, view it on GitHub > <#43 (comment)
> <https://github.com/LibreHealthIO/lh-ehr-laravel/pull/43#issuecomment-1626399181>>,
> > or unsubscribe >
> https://github.com/notifications/unsubscribe-auth/AACFZC324C54XWM3WP4KX4LXPCV2BANCNFSM6AAAAAAZA67IAQ
> > . > You are receiving this because you are subscribed to this
> thread.Message > ID: *@*.***> > -- https://thinkstorm.net 503-330-2239 --
> https://thinkstorm.net 503-330-2239
>
> so as of now can we move forward with a single user name field? I think it
> will be an efficient way to store the name.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LibreHealthIO/lh-ehr-laravel/pull/43#issuecomment-1627391990>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACFZCYVHONB3LULTEXV7ZLXPGBTRANCNFSM6AAAAAAZA67IAQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
https://thinkstorm.net
503-330-2239

muarachmann commented:
Thanks @tmccormi for your inputs, lets separate it at least to `given name` & `surname`.  That is the norm I see other wise we can keep the old format.

chaitak-gorai commented:
> Thanks @tmccormi for your inputs, lets separate it at least to `given name` & `surname`. That is the norm I see other wise we can keep the old format.

ok i ll update it to first name and last name

muarachmann commented:
@chaitak-gorai we need to equally work on the invitation module. This when a user is created should send an email to them to accept the invite

muarachmann commented:
With this we can then focus on the restrictions of the `users -roles / permissions`

chaitak-gorai commented:
> @chaitak-gorai we need to equally work on the invitation module. This when a user is created should send an email to them to accept the invite

yes @muarachmann i am going through it. But cant figure out how to properly implement that. Can we have a call today or tomorrow for a short discussion on this ??

muarachmann commented:
> > @chaitak-gorai we need to equally work on the invitation module. This when a user is created should send an email to them to accept the invite
> 
> yes @muarachmann i am going through it. But cant figure out how to properly implement that. Can we have a call today or tomorrow for a short discussion on this ??

Yes @chaitak-gorai let's have a call tomorrow 1PM GMT + 1

chaitak-gorai commented:
> > > @chaitak-gorai we need to equally work on the invitation module. This when a user is created should send an email to them to accept the invite
> > 
> > 
> > yes @muarachmann i am going through it. But cant figure out how to properly implement that. Can we have a call today or tomorrow for a short discussion on this ??
> 
> Yes @chaitak-gorai let's have a call tomorrow 1PM GMT

okk thank you

chaitak-gorai committed:
➕ Add User via Invitation

chaitak-gorai commented:
## Update: Add User Via Invitation 
@muarachmann I have added the functionality of adding a user via Mail invitation. After admin add a user, the user will get a mail in his email and a link to set a password. There is a token which is valid for 7 days. If it gets expired the user cant add a password. Here are the changes i have made
1. `Add users` form updated for the name fields 
2. Update the User Model and database schema for form fields that were not there and new fields like token, and token_expiry time for verification.
3. Create a Mail class for sending mail.
4. Created the `Add Password` form where the user will provide a new password.
5. Added new routes to web.php

Task Left To be done:
- Form validation request.
- Success and failure messages in the UI.

![image](https://github.com/LibreHealthIO/lh-ehr-laravel/assets/77141674/390a4b4d-a3b3-4745-b6e4-94967c963b77)

chaitak-gorai committed:
Added a  default value for token expiry

muarachmann commented:
Hi @chaitak-gorai ,

This is no where close to what I asked you to do please consider the architecture where invitations are managed separately see - https://github.com/junaidnasir/larainvite

Also please user `mailhog` / `mailpit` for testing emails locally rather than using your gmail. 

regards.
MR

chaitak-gorai commented:
> Hi @chaitak-gorai ,
> 
> This is no where close to what I asked you to do please consider the architecture where invitations are managed separately see - https://github.com/junaidnasir/larainvite
> 
> Also please user `mailhog` / `mailpit` for testing emails locally rather than using your gmail.
> 
> regards. MR

Thank you for correcting me. I went through the suggested code and I got the idea where I am wrong. Will be updating the changes in a new commit.

chaitak-gorai committed:
User Invitation Updated

chaitak-gorai commented:
hey @muarachmann, I have updated the code with a separate invitation class to handle all invitation-related logic.
Also, I have added two tables `user_invitations` which will keep the details about the invitation, and `users_invited` which will keep the user form data that will be used to create the real user when the invite is accepted.
Also, update the login in the Invitation accepting URL. Based on the status of the invitation the URL will be shown. 

Please review whether this approach is correct or not and suggest me changes if required.
After this, I will be working on the dashboard to show all the invitations with their status.

muarachmann commented:
Hi @chaitak-gorai this looks like it, no need for another model `UserInvite` let's create them as `User` model and attach it. Also we can rename the table to `invitations` . Also the invitatation should be facility specific, You can use morphs for this to handle even the `invitee`

chaitak-gorai commented:
@muarachmann yes i have kept the invite in the User Model only in my last approach. 

But I thought it will be better if we keep them separate as the user can reject an invitation or it can expire. In that case, the created user will be null and void. So creating the user only after the invitation is accepted and a password is set .

Should i revert this and keep a single table? I will be updating the suggested changes then.

muarachmann commented:
Yes please keep a single table. If a user rejects, just mark the user as
deleted, inactive

On Wed, Jul 26, 2023 at 1:55 AM Chaitak Gorai ***@***.***>
wrote:

> @muarachmann <https://github.com/muarachmann> yes i have kept the invite
> in the User Model only in my last approach.
>
> But I thought it will be better if we keep them separate as the user can
> reject an invitation or it can expire. In that case, the created user will
> be null and void. So creating the user only after the invitation is
> accepted and a password is set .
>
> Should i revert this and keep a single table? I will be updating the
> suggested changes then.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LibreHealthIO/lh-ehr-laravel/pull/43#issuecomment-1650804296>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4X4XURLBI5437ZOTGVN7TXSBTGNANCNFSM6AAAAAAZA67IAQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

*MUA N. LAURENT*: Head of Engineering
WASPITO <https://waspito.com/>
***@***.***
Bonamoussadi Denver, Douala


*#HealthWithoutAStep*

chaitak-gorai committed:
Table Updates

chaitak-gorai commented:
## Update : 
@muarachmann, I have updated the suggested changes. 
- Drop the `user_invited` table. Only single table for `User`. and renamed the `user_inviations`  into  `Invitation`.
- used `token` instead of `code`

[Help Required] I have used Axios to send some requests where JSON response is needed. I can't figure out how to implement it with `Inertia`. Can you help me with this?

muarachmann commented:
> ## Update :
> @muarachmann, I have updated the suggested changes.
> 
> * Drop the `user_invited` table. Only single table for `User`. and renamed the `user_inviations`  into  `Invitation`.
> * used `token` instead of `code`
> 
> [Help Required] I have used Axios to send some requests where JSON response is needed. I can't figure out how to implement it with `Inertia`. Can you help me with this? 

Why do you need a json response in those routes ? That is the first question you should be asking yourself. You don't need a separate route for the inivitation status, if it has expired show it in the view that it has expired.

With regards to the invitations table.

Emails are unique for these kind of stuffs and we can have the facility that invited the user. What I suggest is using `Morph` relationship to know the `invitee`, `invitor` and `facility` in the invitations table and also that will help isolate the `for_user` used in the users table which has no need. In the end you will have a powerful module that will manage users invitations accross the app. If you are worried about roles and permissions, You are free to store it in the invitations table column say `invite_params` as a json object where when accepted you can assign the user these roles. So a user can be created with basic user access role.

Let me know if you need a call for this
Regards,
MR

chaitak-gorai commented:
> > ## Update :
> > @muarachmann, I have updated the suggested changes.
> > 
> > * Drop the `user_invited` table. Only single table for `User`. and renamed the `user_inviations`  into  `Invitation`.
> > * used `token` instead of `code`
> > 
> > [Help Required] I have used Axios to send some requests where JSON response is needed. I can't figure out how to implement it with `Inertia`. Can you help me with this?
> 
> Why do you need a json response in those routes ? That is the first question you should be asking yourself. You don't need a separate route for the inivitation status, if it has expired show it in the view that it has expired.
> 
> With regards to the invitations table.
> 
> Emails are unique for these kind of stuffs and we can have the facility that invited the user. What I suggest is using `Morph` relationship to know the `invitee`, `invitor` and `facility` in the invitations table and also that will help isolate the `for_user` used in the users table which has no need. In the end you will have a powerful module that will manage users invitations accross the app. If you are worried about roles and permissions, You are free to store it in the invitations table column say `invite_params` as a json object where when accepted you can assign the user these roles. So a user can be created with basic user access role.
> 
> Let me know if you need a call for this Regards, MR



> > ## Update :
> > @muarachmann, I have updated the suggested changes.
> > 
> > * Drop the `user_invited` table. Only single table for `User`. and renamed the `user_inviations`  into  `Invitation`.
> > * used `token` instead of `code`
> > 
> > [Help Required] I have used Axios to send some requests where JSON response is needed. I can't figure out how to implement it with `Inertia`. Can you help me with this?
> 
> Why do you need a json response in those routes ? That is the first question you should be asking yourself. You don't need a separate route for the inivitation status, if it has expired show it in the view that it has expired.
> 
> With regards to the invitations table.
> 
> Emails are unique for these kind of stuffs and we can have the facility that invited the user. What I suggest is using `Morph` relationship to know the `invitee`, `invitor` and `facility` in the invitations table and also that will help isolate the `for_user` used in the users table which has no need. In the end you will have a powerful module that will manage users invitations accross the app. If you are worried about roles and permissions, You are free to store it in the invitations table column say `invite_params` as a json object where when accepted you can assign the user these roles. So a user can be created with basic user access role.
> 
> Let me know if you need a call for this Regards, MR

Yes it will be better if we can have a call. Are you available tomorrow??

chaitak-gorai commented:
> > ## Update :
> > @muarachmann, I have updated the suggested changes.
> > 
> > * Drop the `user_invited` table. Only single table for `User`. and renamed the `user_inviations`  into  `Invitation`.
> > * used `token` instead of `code`
> > 
> > [Help Required] I have used Axios to send some requests where JSON response is needed. I can't figure out how to implement it with `Inertia`. Can you help me with this?
> 
> Why do you need a json response in those routes ? That is the first question you should be asking yourself. You don't need a separate route for the inivitation status, if it has expired show it in the view that it has expired.
> 
> With regards to the invitations table.
> 
> Emails are unique for these kind of stuffs and we can have the facility that invited the user. What I suggest is using `Morph` relationship to know the `invitee`, `invitor` and `facility` in the invitations table and also that will help isolate the `for_user` used in the users table which has no need. In the end you will have a powerful module that will manage users invitations accross the app. If you are worried about roles and permissions, You are free to store it in the invitations table column say `invite_params` as a json object where when accepted you can assign the user these roles. So a user can be created with basic user access role.
> 
> Let me know if you need a call for this Regards, MR

Yes, it will be better to discuss this on a call. Are you free tomorrow ??

muarachmann commented:
Tomorrow 1 pm WAT is fine with me.

On Thu, 27 Jul 2023 at 12:08 Chaitak Gorai ***@***.***> wrote:

> Update :
>
> @muarachmann <https://github.com/muarachmann>, I have updated the
> suggested changes.
>
>    - Drop the user_invited table. Only single table for User. and renamed
>    the user_inviations into Invitation.
>    - used token instead of code
>
> [Help Required] I have used Axios to send some requests where JSON
> response is needed. I can't figure out how to implement it with Inertia.
> Can you help me with this?
>
> Why do you need a json response in those routes ? That is the first
> question you should be asking yourself. You don't need a separate route for
> the inivitation status, if it has expired show it in the view that it has
> expired.
>
> With regards to the invitations table.
>
> Emails are unique for these kind of stuffs and we can have the facility
> that invited the user. What I suggest is using Morph relationship to know
> the invitee, invitor and facility in the invitations table and also that
> will help isolate the for_user used in the users table which has no need.
> In the end you will have a powerful module that will manage users
> invitations accross the app. If you are worried about roles and
> permissions, You are free to store it in the invitations table column say
> invite_params as a json object where when accepted you can assign the
> user these roles. So a user can be created with basic user access role.
>
> Let me know if you need a call for this Regards, MR
>
> Yes, it will be better to discuss this on a call. Are you free tomorrow ??
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LibreHealthIO/lh-ehr-laravel/pull/43#issuecomment-1653399957>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4X4XWTW2UVNSDWEAL46N3XSJD2DANCNFSM6AAAAAAZA67IAQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

chaitak-gorai commented:
> Tomorrow 1 pm WAT is fine with me.
> […](#)
> On Thu, 27 Jul 2023 at 12:08 Chaitak Gorai ***@***.***> wrote: Update : @muarachmann <https://github.com/muarachmann>, I have updated the suggested changes. - Drop the user_invited table. Only single table for User. and renamed the user_inviations into Invitation. - used token instead of code [Help Required] I have used Axios to send some requests where JSON response is needed. I can't figure out how to implement it with Inertia. Can you help me with this? Why do you need a json response in those routes ? That is the first question you should be asking yourself. You don't need a separate route for the inivitation status, if it has expired show it in the view that it has expired. With regards to the invitations table. Emails are unique for these kind of stuffs and we can have the facility that invited the user. What I suggest is using Morph relationship to know the invitee, invitor and facility in the invitations table and also that will help isolate the for_user used in the users table which has no need. In the end you will have a powerful module that will manage users invitations accross the app. If you are worried about roles and permissions, You are free to store it in the invitations table column say invite_params as a json object where when accepted you can assign the user these roles. So a user can be created with basic user access role. Let me know if you need a call for this Regards, MR Yes, it will be better to discuss this on a call. Are you free tomorrow ?? — Reply to this email directly, view it on GitHub <[#43 (comment)](https://github.com/LibreHealthIO/lh-ehr-laravel/pull/43#issuecomment-1653399957)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AD4X4XWTW2UVNSDWEAL46N3XSJD2DANCNFSM6AAAAAAZA67IAQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>
Hey @muarachmann I got some urgent family commitments today. Can it be rescheduled for tomorrow at 1 PM?

chaitak-gorai committed:
Minor Updates

chaitak-gorai commented:
## Update

- Axios calls are removed now passed the invitation status in the view to show a message if expired or accepted.
- Single controller for accepting and rejecting invites based on the reject key.
- Removed irrelevant links between  `users` and `invitations` table. User details can be extracted by email only.
- The URL of the invitation page is changed `/invitations/join/token`
- Confirm Password feature is added when setting up a password.
- Routes names are updated.

muarachmann commented on app/Http/Controllers/Dashboard/User/UserController.php:
```diff
@@ -17,22 +17,13 @@
+         * Shows the role page
+         * @return Response
+         */
+        return Inertia::render('AddUsers');
```
Have this in the users folder `Users/SetupAccount`

muarachmann commented on app/Http/Controllers/Dashboard/User/UserController.php:
```diff
@@ -86,39 +77,38 @@ public function store(Request $request)
+        } else {
 
-          return Redirect::back()->with(['error' => 'Token Mismatch']);
+            return Redirect::back()->with(['error' => 'Token Mismatch']);
```
Just let them know the token is expired

muarachmann commented on app/Http/Controllers/UserInvitationController.php:
```diff
@@ -3,20 +3,46 @@
+        [$username, $domain] = explode('@', $email);
+        $usernameLength = strlen($username);
+        $firstHalf = substr($username, 0, ceil($usernameLength / 2));
+        $secondHalf = str_repeat('*', floor($usernameLength / 2));
```
You can use `Str::mask(....)` to achieve this

muarachmann commented on app/Invite.php:
```diff
@@ -17,38 +18,37 @@ public static function send($userData, $email,)
-            now()->addDays(7),
-            ['username' => $userData['username'],'token'=>$token]
-        );
+        $url = url("/invitations/join/{$token}");
```
Use route name here

muarachmann commented:
@chaitak-gorai cool we are getting there... left some reviews

chaitak-gorai committed:
Minor updates on reviews

chaitak-gorai commented:
@muarachmann I have updated the suggested changes
- Renamed `UserInvitationController` into just `InvitationController`
- Removed redundant code inside `UserController` which is now done inside the `Invite` Class only.
- Used `Str::mask` to create the mask email.
- Used named route while creating the invite URL.
- Added  a separate controller `SetupAccount` for showing `AddUsers` as you asked for

chaitak-gorai committed:
Invitations List UI

chaitak-gorai commented:
## Invitations List UI

Hey @muarachmann, I have created the dashboard UI to show the list of Invitations sent to the Users.
- It includes basic details like `Name`, `Email` of the User, `facility`, `role`, and `status` of the invitaiton.
- Specific font color based on 'accepted', 'rejected', or 'pending'.
- Also added a dropdown to filter data based on the `facility`

here is the Screenshot. Please give it a review and suggest me changes if required.
![image](https://github.com/LibreHealthIO/lh-ehr-laravel/assets/77141674/d5491a85-d433-41f1-b744-36814b64dfa0)

muarachmann commented on app/Http/Controllers/InvitationController.php:
```diff
@@ -12,6 +12,13 @@
+    public function showInvitations(): Response
+    {
+        // Fetch invitation data from the database
+        $invitations = Invitation::with('user')->get();
```
Paginate results here , you expect to have hundreds of invitations listed here. Also any reason why the default datatable was not used ?

chaitak-gorai replied:
Hey @muarachmann, I didn't understand what you mean by default database here. here we are using our invitation table and fetching the user details..

muarachmann replied:
There is vue datatable included in the project use it. See how users are displayed in the table.

muarachmann commented on app/Models/Invitation.php:
```diff
@@ -10,4 +10,8 @@ class Invitation extends Model
     protected $table = 'invitations';
     protected $fillable = ['email', 'facility', 'token', 'status', 'valid_till', 'accepted_at'];
+    public function user()
+    {
```
spacing in coding

muarachmann commented on resources/app/pages/Invitations.vue:
```diff
@@ -0,0 +1,130 @@
+    },
+    data() {
+        return {
+            selectedFacility: "", // Holds the selected facility for filtering
```
Could be null for all facilities.

muarachmann commented on resources/app/pages/Invitations.vue:
```diff
@@ -0,0 +1,130 @@
+        uniqueFacilities() {
+            console.log(this.invitations);
+            return [
+                ...new Set(
```
This filter is suppose to be done backend.

chaitak-gorai committed:
Updates with Pagination.

chaitak-gorai commented:
## Update: Use of Datatable, Pagination, Filtering
@muarachmann, I have updated the reviewed changes. 
- Used the `data-table` to show the invitations table.
- Used Pagination, Filtering by `facility` on the server side.
- Updated the views and passed the facilities' names to show in the dropdown.

Please review and suggest to me if further updates are required.

muarachmann commented:
@chaitak-gorai good job... good progress so far. Now ensure everything in invitations is under the users folder (frontend).
Also ensure you extend the layouts from the dashboard layout and the accept/reject from the auth layout. With that just a few polishing we can work on the headers and access by roles now

chaitak-gorai committed:
Proper Folder Structure Update

chaitak-gorai commented:
Thank You @muarachmann for the reviews. I have a great learning curve so far in this project. 

- I have added the frontend pages for Invitaitons inside `Users` Folder and Roles and Permissions related pages inside `AccessControl` Folder. 
- Also added the `Dashboard Layout` and `Auth Layout` for the respective pages

chaitak-gorai committed:
User Profile Page

chaitak-gorai commented:
## Update
- Create a `User Profile` Page with the current details from the `Add-User` form. As more fields are added we can update it.
- Linked this in the invitations list for each User.

This is how it looks:
![image](https://github.com/LibreHealthIO/lh-ehr-laravel/assets/77141674/0581b5de-ffda-45d5-901a-f68a74dd68ae)

muarachmann commented:
> ## Update
> * Create a `User Profile` Page with the current details from the `Add-User` form. As more fields are added we can update it.
> * Linked this in the invitations list for each User.
> 
> This is how it looks: ![image](https://user-images.githubusercontent.com/77141674/260331325-0581b5de-ffda-45d5-901a-f68a74dd68ae.png)

Nice @chaitak-gorai. Please add buttons to edit profile and roles. Also add breadcrumbs like the other screens so it looks better.

chaitak-gorai commented:
Hey @muarachmann i have a query about the value provided in the active link of Breadcrumb component as like in dashboard page it is set to 
`activeLink: i18n.t('general.dashboard'),`
how it is set like we just give a name of the page or it is dynamically calculated? I have a confusion here.
Similarly for the users page will we just pass "userProfile" in this function?

chaitak-gorai committed:
Edit Profile Functionality

chaitak-gorai committed:
Minor update

muarachmann commented on app/Http/Controllers/Dashboard/User/UserController.php:
```diff
@@ -69,15 +70,27 @@ public function profile($userId): Response
         return Inertia::render('Users/UserProfile', ['userData' => $user]);
     }
 
+    public function showEditPage($userId): Response
```
You can just call this method `edit`

chaitak-gorai commented:
## Update: Edit User Functionality 
- Added a `Edit User` button on top right of User Profile Page.
- It will take to the Edit Usre Page.
- Made a copy of  `Add User ` form and Created a Edit User form from it which will show the user data and roles.
- On submit the User data is updated and if roles are updated too then old roles and permission are revoked and new are added

![image](https://github.com/LibreHealthIO/lh-ehr-laravel/assets/77141674/88f59f31-816a-4cd2-ac78-a72954ab1c1b)

muarachmann commented on app/Http/Controllers/Dashboard/User/UserController.php:
```diff
@@ -69,15 +70,27 @@ public function profile($userId): Response
+        $user = User::where('id', $userId)->first();
+        $facilities = Facility::pluck('name', 'id');
+
+        return Inertia::render('Users/EditUser', ['userData' => $user, 'facilities' => $facilities]);
```
Same here `Users/Edit`

muarachmann commented on app/Http/Controllers/Dashboard/User/UserController.php:
```diff
@@ -69,15 +70,27 @@ public function profile($userId): Response
      * @return Response
      */
-    public function edit($id)
+    public function edit()
```
Here you have the method, so use it

muarachmann commented on resources/app/pages/Users/UserProfile.vue:
```diff
@@ -156,14 +164,24 @@
     data() {
-        return {};
+        return {
+            activeLink: i18n.t("userProfile"),
```
use snake case here please `user_profile`

muarachmann commented on resources/app/pages/Users/UserProfile.vue:
```diff
@@ -156,14 +164,24 @@
+    },
+    methods: {
+        editLink() {
+            console.log(this.userData.id);
```
This can be done in a link tag `<inertia :href....`

muarachmann commented on routes/web.php:
```diff
@@ -129,15 +129,15 @@ function () {
             Route::post('/add-user', [InvitationController::class, 'sendInvite'])->name('users.add');
             Route::get('/invitations', [InvitationController::class, 'showInvitations'])->name('user.invitations');
             Route::get('/users/profile/{userId}', [UserController::class, 'profile'])->name('users.profile');
+            Route::get('/users/edit/{userId}', [UserController::class, 'showEditPage'])->name('users.edit');
```
Please use the resource routes for these

muarachmann commented on routes/web.php:
```diff
@@ -124,13 +124,6 @@ function () {
-            Route::resource('users', UserController::class)->names([
-                'index' => 'users.index',
-                'create' => 'users.create',
-                'show' => 'users.show',
```
This should remain pls

muarachmann commented:
> Hey @muarachmann i have a query about the value provided in the active link of Breadcrumb component as like in dashboard page it is set to `activeLink: i18n.t('general.dashboard'),` how it is set like we just give a name of the page or it is dynamically calculated? I have a confusion here. Similarly for the users page will we just pass "userProfile" in this function?

Good question, look at the patient profile component in the codebase.

chaitak-gorai commented:
Thank you for the reviews. I will update them ASAP.

chaitak-gorai committed:
Routes Update, Clean file names

chaitak-gorai committed:
Removed redundant file

chaitak-gorai commented:
## Update: 
- Used the resourse routes instead to custom routes for `users` related functions.
   - Remove the custom routes and change the named routes everywhere it is used.
   - Moved the user addition logic from Invitations controller inside `store` functions of user controller.
   - Removed redundant `Setup Account` controller. It can be done inside `create` of User Controller.
- Added Breadcrumbs in all User Pages `Users/Invitations->Full Name->Edit`  or  `Users/Invitations->Add`
- Used Inertia link as suggested in review.
- Rename file names with clean practices 
   - `AddUsers.vue` into `Users/Add.vue`
   - `EditUsers.vue` into `Users/Edit.vue`
   - `UsersProfile.vue` into `Users/Profile.vue`

@muarachmann these are the bunch of changes I have updated based on your reviews. Please give it a review and suggest how to proceed further.

chaitak-gorai committed:
multiselect component for permissions select

chaitak-gorai commented:
@muarachmann i have added a multiselect input type in `EhrInput` and replaced the old select dropdown for permission with this component in `Roles.vue` page.
Please review this and suggest me changes if required.

muarachmann commented:
Hi @chaitak-gorai please use this - MultiSelect component in the project

chaitak-gorai commented:
> Hi @chaitak-gorai please use this - MultiSelect component in the project

I looked for the component in the project but couldn't able to find it. Can you please tell me where it is located. Do you mean by the `vue-multiselect` package?

chaitak-gorai committed:
Revert last changes and used MultiSelect

muarachmann commented:
@chaitak-gorai do not forget to list the users in the datatable as well.
Regards

chaitak-gorai committed:
Merge branch 'develop' into roles

chaitak-gorai committed:
access controlled routes

chaitak-gorai committed:
Merge branch 'roles' of https://github.com/chaitak-gorai/lh-ehr-laravel into roles

chaitak-gorai committed:
Users datatable

chaitak-gorai commented:
## Update

@muarachmann  i have added these changes
- Rebase `header-fix` branch for the updated menu link.s
- Added Users Datatable 
- Added links to the pages in `Administration` menu link for `Invitations`, `Users` and `ACL` for roles.
- modified the routes for roles and permission-based accessing. As discussed in the call, we don't have to create separate middleware for that. I used the default middleware of `Laratrust` where we can use both permission and role-based access for the routes.

I haven't hide the menu link in frontend. It will be better if we show them a `Acess Denied` Page instead .  Any thought?

muarachmann commented:
Hi @chaitak-gorai will be reviewing this this weekend for a merge. Anything you have to polish?

chaitak-gorai commented:
> Hi @chaitak-gorai will be reviewing this this weekend for a merge. Anything you have to polish?

Hey @muarachmann, thank you for the update. I'll look into it and do the necessary polishing by the weekend. There are some validations left i guess. I'll update the changes if any. 
Regards,
Chaitak Gorai

chaitak-gorai committed:
Validations for Add User

chaitak-gorai commented:
Hey @muarachmann, i have added the validation for the `Add-User` form which was left. As of now, i have done with polishing the PR. You can suggest me any further polishing if required.  Thank you.

chaitak-gorai commented:
Hey @muarachmann hope you are doing fine. 
I just want to remind you when it will be merged? Is there any changes i need to commit?
Thank you.

muarachmann commented:
> Hey @muarachmann hope you are doing fine. I just want to remind you when it will be merged? Is there any changes i need to commit? Thank you.

Hi @chaitak-gorai I have been really busy lately. Will review this during the weekend and merge

chaitak-gorai commented:
Hey @muarachmann its ok if you are busy. Just a reminder that if you are available this weekend can we discuss on this.
Also, can we also discuss about the other modules which we can build in upcoming session.
